### PR TITLE
Support tctl operations with the ClusterConfig resource

### PR DESCRIPTION
**Pull request by andrejtokarcik**
The purpose of the current PR is to partially address #4867; RBAC tests are not covered here.

The following `tctl` resource-manipulation operations are now supported with `ClusterConfig`:

### `tctl get cluster_config`

With `--format=text`, I made the judgment call to choose the columns `Cluster ID`, `Session Recording` and `Proxy Checks Host Keys?` to be displayed in the table, let me know if I should change that.

### `tctl create -f cc.yaml`
This overwrites the current cluster config as a whole, fields not specified in the YAML are reset to their default value.

  - **Warning**: The reset-to-default behaviour does not work for booleans. The unmarshalling machinery in general does not seem to play well with boolean-valued fields, since it apparently does not distinguish between `false` when given explicitly in the YAML as opposed to when supplied in place of a missing value.

    In the case of `ClusterConfig` this causes the `local_auth` field to be set to `false` when not specified even though its default value ought to be `true`. There is a way to fix this using the `services.DefaultClusterConfig` function but a better approach would probably be to resolve this more generally.

Without `-f` one gets `error: cluster configuration can be only overwritten or updated, use -f or --force flag to overwrite`.

### `tctl update cluster_config`

The auth server's gRPC API is extended with an `UpdateClusterConfig` method, similar to `UpdateRemoteCluster`.

The only available CLI option is currently `--set-session-recording`. What other `--set-*` options should `tctl update cluster_config` provide, i.e. which other `ClusterConfig` fields should be updatable? I still don't understand how the `ClusterConfig` resource is supposed to allow for changing the 2FA settings -- the resource doesn't appear to involve any such field.

Also note that the `--set-labels`/`--set-ttl` flags meant for `RemoteCluster` can still be used when updating `ClusterConfig` (provided `--set-session-recording` is present): the `RemoteCluster`-specific options would then have no effect but may end up confusing from the user's perspective. Ultimately, for `tctl update` to work with YAML files (consisting of whitelisted fields) is probably the way to go.

### CODE REVIEW

**Review by klizhentas — CHANGES_REQUESTED** [review]
I think `--set-labels` and `--set-ttl` are OK as separate flags, because they are common for every resource.

The flag `--set-recording-mode` as a separate flag will be unclear, because not all resources have it.

Instead, I suggest `tctl update -f resource.yaml` behave like [kubectl patch](https://kubernetes.io/docs/tasks/manage-kubernetes-objects/update-api-object-kubectl-patch/)

This will be more scalable because it work for every resource. 

@awly @andrejtokarcik @russjones @fspmarshall what do you folks think?

**Comment by fspmarshall** [comment]
> Instead, I suggest `tctl update -f resource.yaml` behave like [kubectl patch](https://kubernetes.io/docs/tasks/manage-kubernetes-objects/update-api-object-kubectl-patch/)

IMO if something behaves like a patch, that should be very explicit (e.g. `tctl update --patch=...`).  Otherwise, you might think you've removed a field, when in fact you simply left it as-is because it wasn't part of the "patch".

---

General question about this PR, won't any changes made by `tctl` in this way get overwritten the next time an auth server joins or restarts?  IIUC auth servers always assume that their filesystem config dictates the most up-to-date cluster configuration and unilaterally overwrite the current cluster configuration as part of their startup logic (which is why updating a config file and then restarting the auth server works).

@klizhentas ^

**Comment by andrejtokarcik** [comment]
@fspmarshall You're correct: as it stands, the changes made to the cluster config via `tctl` won't survive an auth server restart.

@klizhentas The idea (suggested by @russjones) is that not all `--set-*` options are meant to be used with all resource kinds, see my note about how `--set-ttl` and `--set-labels` have no effect when updating `ClusterConfig`.

I agree that passing YAMLs to `tctl update` would be more scalable, though as I mentioned it may require a larger refactor and more thoughtful decisions going beyond the present `ClusterConfig` concerns. Just off the top of my head: (1) the issue of not being able to distinguish the cases of fields not-specified vs. explitictly-set-to-default must be resolved (cf. the Warning above); (2) the patch semantics intended for composite values is not self-evident, `kubectl` provides several merging strategies for that reason.

BTW, what do you mean by `tctl update -f`? Perhaps you're using `-f` to indicate a following file argument? In `tctl create` the flag means `--force` though.

**Comment by awly** [comment]
How about a generic `--set=field.path=value`, e.g. `--set=spec.session_recording=true`?

File: tool/tctl/common/resource_command.go [hunk 1]
```diff
@@ -376,6 +389,34 @@ func (rc *ResourceCommand) createUser(client auth.ClientI, raw services.UnknownR
...
+		return trace.Wrap(err)
+	}
+
+	newConfig, err := services.GetClusterConfigMarshaler().Unmarshal(raw.Raw)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	if newConfig.GetName() != existingConfig.GetName() {
+		return trace.CompareFailed("cluster config resource must have the name %q", existingConfig.GetName())
```
Comment (a-palchikov) [inline tool/tctl/common/resource_command.go:410]: error messages should be consistent with the resource type (`cluster configuration`)

**Review by a-palchikov — COMMENTED** [review]
I also think something more scalable than explicit `set-attr=value` is necessary here. While the generic `set=spec.attr=value` might be a good start and it looks like a YAML patch for a single attribute, I wonder if this suffices for bigger structural changes - e.g. it won't be able to support removing bits of configuration.

**Review by fspmarshall — CHANGES_REQUESTED** [review]
> as it stands, the changes made to the cluster config via `tctl` won't survive an auth server restart.

This should be considered a hard blocker for allowing modification of cluster configuration via `tctl`.  This feature is not useful if it allows making changes that might be silently reverted at any moment.

@klizhentas made this comment as part of #4867:

> If there are no values set in config files, dynamic configuration will not be overwritten with defaults. Otherwise, teleport config file overrides the cluster configuration.

Intelligently ensuring that auth servers don't overwrite non-default values is a good start.  However, I think we need to make a hard distinction between static and dynamic configuration, and give precedence to static configuration 100% of the time.  `tctl` should not be capable of even *temporarily* overruling static configuration.  IMO this is a security issue as much as it is a reliability/usability issue.

**Comment by andrejtokarcik** [comment]
Will be largely revised after #5147 is settled. It also seems that only a subset of the `ClusterConfig` fields should be modifiable via `tctl`.

**Pull request closed without merging**